// minterjs-util/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "COIN_MAX_AMOUNT", {
  enumerable: true,
  get: function get() {
    return _variables.COIN_MAX_AMOUNT;
  }
});
Object.defineProperty(exports, "COIN_MAX_MAX_SUPPLY", {
  enumerable: true,
  get: function get() {
    return _variables.COIN_MAX_MAX_SUPPLY;
  }
});
Object.defineProperty(exports, "COIN_MIN_MAX_SUPPLY", {
  enumerable: true,
  get: function get() {
    return _variables.COIN_MIN_MAX_SUPPLY;
  }
});
Object.defineProperty(exports, "FeePrice", {
  enumerable: true,
  get: function get() {
    return _fee.FeePrice;
  }
});
Object.defineProperty(exports, "TX_TYPE", {
  enumerable: true,
  get: function get() {
    return _txTypes.TX_TYPE;
  }
});
Object.defineProperty(exports, "addressToString", {
  enumerable: true,
  get: function get() {
    return _prefix.addressToString;
  }
});
Object.defineProperty(exports, "bufferToCoin", {
  enumerable: true,
  get: function get() {
    return _coinSymbol.bufferToCoin;
  }
});
Object.defineProperty(exports, "buyCoin", {
  enumerable: true,
  get: function get() {
    return _coinMath.buyCoin;
  }
});
Object.defineProperty(exports, "buyCoinByCoin", {
  enumerable: true,
  get: function get() {
    return _coinMath.buyCoinByCoin;
  }
});
Object.defineProperty(exports, "checkToString", {
  enumerable: true,
  get: function get() {
    return _prefix.checkToString;
  }
});
Object.defineProperty(exports, "coinToBuffer", {
  enumerable: true,
  get: function get() {
    return _coinSymbol.coinToBuffer;
  }
});
Object.defineProperty(exports, "convert", {
  enumerable: true,
  get: function get() {
    return _converter.convert;
  }
});
Object.defineProperty(exports, "convertFromPip", {
  enumerable: true,
  get: function get() {
    return _converter.convertFromPip;
  }
});
Object.defineProperty(exports, "convertToPip", {
  enumerable: true,
  get: function get() {
    return _converter.convertToPip;
  }
});
Object.defineProperty(exports, "isMinterPrefixed", {
  enumerable: true,
  get: function get() {
    return _prefix.isMinterPrefixed;
  }
});
Object.defineProperty(exports, "isValidAddress", {
  enumerable: true,
  get: function get() {
    return _prefix.isValidAddress;
  }
});
Object.defineProperty(exports, "isValidCheck", {
  enumerable: true,
  get: function get() {
    return _prefix.isValidCheck;
  }
});
Object.defineProperty(exports, "isValidPublic", {
  enumerable: true,
  get: function get() {
    return _public.isValidPublic;
  }
});
Object.defineProperty(exports, "isValidPublicKeyString", {
  enumerable: true,
  get: function get() {
    return _prefix.isValidPublicKeyString;
  }
});
Object.defineProperty(exports, "isValidTransaction", {
  enumerable: true,
  get: function get() {
    return _prefix.isValidTransaction;
  }
});
Object.defineProperty(exports, "mPrefixStrip", {
  enumerable: true,
  get: function get() {
    return _prefix.mPrefixStrip;
  }
});
Object.defineProperty(exports, "mPrefixToHex", {
  enumerable: true,
  get: function get() {
    return _prefix.mPrefixToHex;
  }
});
Object.defineProperty(exports, "mToBuffer", {
  enumerable: true,
  get: function get() {
    return _prefix.mToBuffer;
  }
});
Object.defineProperty(exports, "normalizeTxType", {
  enumerable: true,
  get: function get() {
    return _txTypes.normalizeTxType;
  }
});
Object.defineProperty(exports, "numberToBig", {
  enumerable: true,
  get: function get() {
    return _converter.numberToBig;
  }
});
Object.defineProperty(exports, "privateToAddressString", {
  enumerable: true,
  get: function get() {
    return _prefix.privateToAddressString;
  }
});
Object.defineProperty(exports, "publicToAddress", {
  enumerable: true,
  get: function get() {
    return _public.publicToAddress;
  }
});
Object.defineProperty(exports, "publicToString", {
  enumerable: true,
  get: function get() {
    return _public.publicToString;
  }
});
Object.defineProperty(exports, "sellCoin", {
  enumerable: true,
  get: function get() {
    return _coinMath.sellCoin;
  }
});
Object.defineProperty(exports, "sellCoinByBip", {
  enumerable: true,
  get: function get() {
    return _coinMath.sellCoinByBip;
  }
});
Object.defineProperty(exports, "toBuffer", {
  enumerable: true,
  get: function get() {
    return _prefix.toBuffer;
  }
});
Object.defineProperty(exports, "txTypeList", {
  enumerable: true,
  get: function get() {
    return _txTypes.txTypeList;
  }
});

var _public = require("./public.js");

var _converter = require("./converter.js");

var _prefix = require("./prefix.js");

var _fee = require("./fee.js");

var _coinMath = require("./coin-math.js");

var _coinSymbol = require("./coin-symbol.js");

var _txTypes = require("./tx-types.js");

var _variables = require("./variables.js");
