// minter-js-sdk/dist/cjs/tx-data/convert-sell.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {number|string} coinToSell - coin id
 * @param {number|string} coinToBuy - coin id
 * @param {number|string} valueToSell
 * @param {number|string} [minimumValueToBuy=0]
 * @constructor
 */

function SellTxData(_ref) {
  var coinToSell = _ref.coinToSell,
      coinToBuy = _ref.coinToBuy,
      valueToSell = _ref.valueToSell,
      _ref$minimumValueToBu = _ref.minimumValueToBuy,
      minimumValueToBuy = _ref$minimumValueToBu === void 0 ? 0 : _ref$minimumValueToBu;
  utils.validateUint(coinToSell, 'coinToSell');
  utils.validateUint(coinToBuy, 'coinToBuy');
  utils.validateAmount(valueToSell, 'valueToSell');
  utils.validateAmount(minimumValueToBuy, 'minimumValueToBuy');
  this.coinToSell = coinToSell;
  this.coinToBuy = coinToBuy;
  this.valueToSell = valueToSell;
  this.minimumValueToBuy = minimumValueToBuy;
  this.txData = new minterjsTx.TxDataSell({
    coinToSell: utils.integerToHexString(coinToSell),
    coinToBuy: utils.integerToHexString(coinToBuy),
    valueToSell: "0x".concat(minterjsUtil.convertToPip(valueToSell, 'hex')),
    minimumValueToBuy: "0x".concat(minterjsUtil.convertToPip(minimumValueToBuy, 'hex'))
  });
  utils.proxyNestedTxData(this);
}
/**
 * @param {Buffer|string} coinToSell
 * @param {Buffer|string} valueToSell
 * @param {Buffer|string} coinToBuy
 * @param {Buffer|string} minimumValueToBuy
 * @return {SellTxData}
 */

SellTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var coinToSell = _ref2.coinToSell,
      valueToSell = _ref2.valueToSell,
      coinToBuy = _ref2.coinToBuy,
      minimumValueToBuy = _ref2.minimumValueToBuy;

  if (!valueToSell && valueToSell !== 0) {
    throw new Error('Invalid valueToSell');
  }

  return new SellTxData({
    coinToSell: utils.bufferToInteger(minterjsUtil.toBuffer(coinToSell)),
    coinToBuy: utils.bufferToInteger(minterjsUtil.toBuffer(coinToBuy)),
    valueToSell: minterjsUtil.convertFromPip(utils.bufferToInteger(minterjsUtil.toBuffer(valueToSell))),
    minimumValueToBuy: minterjsUtil.convertFromPip(utils.bufferToInteger(minterjsUtil.toBuffer(minimumValueToBuy)))
  });
};
/**
 * @param {Buffer|string} data
 * @return {SellTxData}
 */


SellTxData.fromRlp = function fromRlp(data) {
  return SellTxData.fromBufferFields(new minterjsTx.TxDataSell(data));
};

module.exports = SellTxData;
