// minter-js-sdk/dist/cjs/tx-data/pool-sell-all.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {Array<number|string>} coins - list of coin id
 * @param {number|string} [minimumValueToBuy=0]
 * @constructor
 */

function SellAllPoolTxData(_ref) {
  var coins = _ref.coins,
      _ref$minimumValueToBu = _ref.minimumValueToBuy,
      minimumValueToBuy = _ref$minimumValueToBu === void 0 ? 0 : _ref$minimumValueToBu;
  utils.validateUintArray(coins, 'coins');
  utils.validateAmount(minimumValueToBuy, 'minimumValueToBuy');
  this.coins = coins;
  this.minimumValueToBuy = minimumValueToBuy;
  this.txData = new minterjsTx.TxDataSellAllSwapPool({
    coins: coins.map(function (coin) {
      return utils.integerToHexString(coin);
    }),
    minimumValueToBuy: "0x".concat(minterjsUtil.convertToPip(minimumValueToBuy, 'hex'))
  });
  utils.proxyNestedTxData(this);
}
/**
 * @param {Array<Buffer>} coins
 * @param {Buffer|string} minimumValueToBuy
 * @return {SellAllPoolTxData}
 */

SellAllPoolTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var coins = _ref2.coins,
      minimumValueToBuy = _ref2.minimumValueToBuy;
  return new SellAllPoolTxData({
    coins: coins.map(function (item) {
      return utils.bufferToInteger(item);
    }),
    minimumValueToBuy: minterjsUtil.convertFromPip(utils.bufferToInteger(minterjsUtil.toBuffer(minimumValueToBuy)))
  });
};
/**
 * @param {Buffer|string} data
 * @return {SellAllPoolTxData}
 */


SellAllPoolTxData.fromRlp = function fromRlp(data) {
  return SellAllPoolTxData.fromBufferFields(new minterjsTx.TxDataSellAllSwapPool(data));
};

module.exports = SellAllPoolTxData;
