// minter-js-sdk/dist/cjs/tx-data/send.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 *
 * @param {string} to
 * @param {number|string} value
 * @param {number|string} coin - coin id
 * @constructor
 */

function SendTxData(_ref) {
  var to = _ref.to,
      _ref$value = _ref.value,
      value = _ref$value === void 0 ? 0 : _ref$value,
      coin = _ref.coin;
  utils.validateAddress(to, 'to');
  utils.validateUint(coin, 'coin');
  utils.validateAmount(value, 'value');
  this.to = to;
  this.value = value;
  this.coin = coin;
  this.txData = new minterjsTx.TxDataSend({
    to: minterjsUtil.toBuffer(to),
    coin: utils.integerToHexString(coin),
    value: "0x".concat(minterjsUtil.convertToPip(value, 'hex'))
  });
  utils.proxyNestedTxData(this);
}
/**
 * @param {Buffer|string} to
 * @param {Buffer|string|number} value
 * @param {Buffer|string|number} coin
 * @return {SendTxData}
 */

SendTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var to = _ref2.to,
      value = _ref2.value,
      coin = _ref2.coin;
  return new SendTxData({
    to: minterjsUtil.addressToString(to),
    coin: utils.bufferToInteger(minterjsUtil.toBuffer(coin)),
    value: minterjsUtil.convertFromPip(utils.bufferToInteger(minterjsUtil.toBuffer(value)))
  });
};
/**
 * @param {Buffer|string} data
 * @return {SendTxData}
 */


SendTxData.fromRlp = function fromRlp(data) {
  return SendTxData.fromBufferFields(new minterjsTx.TxDataSend(data));
};

module.exports = SendTxData;
