// minter-js-sdk/dist/cjs/tx-data/vote-update.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {string} version
 * @param {string} publicKey
 * @param {number|string} height
 * @constructor
 */

function VoteUpdateTxData(_ref) {
  var version = _ref.version,
      publicKey = _ref.publicKey,
      height = _ref.height;
  utils.validatePublicKey(publicKey, 'publicKey');
  utils.validateUint(height, 'height');
  this.version = version;
  this.publicKey = publicKey;
  this.height = height;
  this.txData = new minterjsTx.TxDataVoteUpdate({
    version: Buffer.from(version.toString(), 'utf-8'),
    publicKey: minterjsUtil.toBuffer(publicKey),
    height: utils.integerToHexString(height)
  });
  utils.proxyNestedTxData(this);
}
/**
 * @param {Buffer|string} version
 * @param {Buffer|string} publicKey
 * @param {Buffer|string|number} height
 * @return {VoteUpdateTxData}
 */

VoteUpdateTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var version = _ref2.version,
      publicKey = _ref2.publicKey,
      height = _ref2.height;
  return new VoteUpdateTxData({
    version: minterjsUtil.toBuffer(version).toString('utf-8'),
    publicKey: minterjsUtil.publicToString(publicKey),
    height: utils.bufferToInteger(minterjsUtil.toBuffer(height))
  });
};
/**
 * @param {Buffer|string} data
 * @return {VoteUpdateTxData}
 */


VoteUpdateTxData.fromRlp = function fromRlp(data) {
  return VoteUpdateTxData.fromBufferFields(new minterjsTx.TxDataVoteUpdate(data));
};

module.exports = VoteUpdateTxData;
